### PR TITLE
Require Julia 1.6+, use Preferences, eliminate mutable state

### PR DESCRIPTION
**Pull request by staticfloat**
This adds support for `Preferences.jl` as the backing store for setting
a computation backend (currently only supports `fftw` and `mkl` as valid
choices).  The convenience method `set_provider!()` is given, but it is
little more than a wrapper around `Preferences.set_preferences!()`.
Users can alter their `LocalPreferences.toml` or `Project.toml` files by
hand as well; in the event that an illegal value is provided, an error
will be displayed and `fftw` will be chosen as the default value.

Preferences is a relatively new subsystem in Julia and it has some
lacking support in `Pkg`; so for now we are adding a separate
`test/Project.toml`, which will have preferences added into it on CI, as
there is a bug in propagating preferences when a single `Project.toml`
is used with extra dependencies for the tests.

### CODE REVIEW

**Comment by staticfloat** [comment]
This closes https://github.com/JuliaMath/FFTW.jl/issues/188

X-ref: https://github.com/JuliaLang/Pkg.jl/issues/2500

**Review by stevengj — APPROVED** [review]

**Comment by stevengj** [comment]
LGTM, assuming CI is okay…

**Comment by giordano** [comment]
@staticfloat perhaps it'd also be useful to have some documentation about how to set the preference :slightly_smiling_face:

File: src/providers.jl [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        end
+    end
+end
+
+# If we're using MKL, load it in and set library paths appropriately.
+@static if fftw_provider == "mkl"
+    using IntelOpenMP_jll, MKL_jll
+    const libfftw3 = MKL_jll.libmkl_rt_path
+    const libfftw3f = libfftw3
+end
```
Comment (andreasnoack) [inline src/providers.jl:80]: This doesn't seem compatible with PackageCompiler. The paths will end up being compiled into the system image.

Comment (staticfloat) [reply src/providers.jl:80]: Correct, this should be done within `__init__()` if we need this to work with PackageCompiler.

Comment (andreasnoack) [reply src/providers.jl:80]: Generally we ned all packages to work with PackageCompiler and specifically we very much need this to work with PackageCompiler

**Pull request merged**